### DISCUSSION TITLE
Need to call one config api but only first time on serverside

### DISCUSSION THREAD

**Discussion by rajatdhoot123** [vercel/next.js#34458] [Help]
Is there any way we can call API only the first time while loading the side in serverside...?.

It's like calling getServerSide props in _app

**Comment by 2stash** [comment]
@rajatdhoot123 Hello! I'm. not sure I understand your question, but you can call getServerSideProps from your page files. So yes you can call an api from getServerSideProps that runs on the server. There is nothing in the api docs that has a config that you can specify that it runs once. But you can put some caching logic so that although the function will be called many times the logic will only once; similar to how you can configure useEffect in React.

https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props

The other option is to just use the getStaticProps and pass 'false' to the 'revalidate' config options. This will only run the getStaticProps function once on loading. Now I think there are some caveats I don't get yet, but this code also runs server side, but is intended for data that doesn't change often.

https://nextjs.org/docs/basic-features/data-fetching/get-static-props

Here is a good video/blog post of the differences:
https://medium.com/eincode/next-js-data-fetching-getstaticprops-vs-getserversideprops-fcbf43d0ccac

**Reply by rajatdhoot123** [reply]
Hey @2stash Thanks for the detailed answer but here I am looking for something else for the client-side we can maintain some states per user but on the server need to check if the API is called by some other page serverSideProp and data is already cached on the client so no need to call again in other page serverSideProps.

Example There are 3 pages -> A, B, C user can open any page directly say user opens B at the first time not I want to fetch user details because it was the first open page and details are not stored on client-side not I call the user details API on getServerSide props of page B now user navigates to page A here no need to call user details API because user details are already on client side because we fetched on page B getServerSideProps. The check weather to call or not how we can maintain at different pages getServerSideProps.

**Reply by 2stash** [reply]
@rajatdhoot123 So I think you can just use normal React state management with redux. See example repo: https://github.com/2stash/nextjs-examples/tree/main/nextjs-login-state

There is a sidebar layout you can toggle a "login" and each page will display the login status. Redux passes the state to needed components. Obviously need to expand on it but it shows how you can share state. If you really want to use getServerSideProps, just save the first getServerSideProps data into redux state, and have a check in getServerSideProps if state is already populated and you can use a simple if to bypass the get data logic.

The official examples also have one that uses typescript:

https://github.com/vercel/next.js/tree/canary/examples/with-redux

**Reply by rajatdhoot123** [reply]
@2stash I don't think there is a way to get the client-side state in getServerSideProps, yes but you can update the client-side state by passing data from which we get from getServerSideProps.

I checked the above two examples but was not able to find the solution for my usecase.

**Reply by 2stash** [reply]
@rajatdhoot123  Good morning again! You are right, not sure what I was thinking, it was late lol.  So I think you have a few options. first is to use an auth package. 

https://nextjs.org/docs/authentication

In the example, it checks for !user and redirects, but I think you could do if(user){don't fetch new data}. Although the getServerSideProps function will run.

The second thing you might be able to do is to not use getServerSideProps. One of the main benefits of getServerSideProps is having your data in the initial HTML so you have good SEO, etc. However with logged in user data this isn't really needed. So, not knowing your full use case, wouldn't it be simpler to use normal react client login state for this data?

**Reply by rajatdhoot123** [reply]
@2stash Use case is for the serverSide to get the SEO benefits yes but will look to the above example which uses the session technique for user determination.

Thanks.
